src/Microsoft.Extensions.Configuration.UserSecrets/ConfigurationExtensions.cs

```diff
@@ -24,9 +29,53 @@ public static IConfigurationBuilder AddUserSecrets(this IConfigurationBuilder co
                 throw new ArgumentNullException(nameof(configuration));
             }
 
-            var fileProvider = configuration.GetFileProvider();
+            var entryAssembly = Assembly.GetEntryAssembly();
+            var attribute = entryAssembly.GetCustomAttribute<UserSecretsIdAttribute>();
+            if (attribute != null)
+            {
+                return AddUserSecrets(configuration, attribute.UserSecretsId);
+            }
+
+            // try fallback to project.json for legacy support
+            try
+            {
+                var fileProvider = configuration.GetFileProvider();
+#pragma warning disable CS0618
+                return AddSecretsFile(configuration, PathHelper.GetSecretsPath(fileProvider));
+#pragma warning restore CS0618
+            }
+            catch
+            { }
+
+            // show the error about missing UserSecretIdAttribute instead of problems with project.json
+            throw new InvalidOperationException(Resources.FormatError_Missing_UserSecretsIdAttribute(entryAssembly.FullName));
+        }
+
+        /// <summary>
+        /// Adds the user secrets configuration source.
+        /// </summary>
+        /// <param name="configuration"></param>
+        /// <param name="assembly">The assembly with the <see cref="UserSecretsIdAttribute" /></param>
+        /// <returns></returns>
+        public static IConfigurationBuilder AddUserSecrets(this IConfigurationBuilder configuration, Assembly assembly)
+        {
+            if (configuration == null)
+            {
+                throw new ArgumentNullException(nameof(configuration));
+            }
+
+            if (assembly == null)
+            {
+                throw new ArgumentNullException(nameof(assembly));
+            }
+
+            var attribute = assembly.GetCustomAttribute<UserSecretsIdAttribute>();
```

Comment:
Sure, could be done, but as I was implementing it didn't seem like it would make it any cleaner.